File: web/client/src/features/favoriteArticles/repository/favoriteArticle.ts
"use server";
// eslint-disable-next-line import/named
import { v4 as uuidv4 } from "uuid";

import { createGetOnlyServerSideClient } from "@/lib/supabase/client/serverClient";

import { Database } from "@/types/database.types";
import { FavoriteArticleType } from "@/types/favoriteArticle";

const LIMIT = 20;

type GetFavoriteArticlesByFavoriteArticleFolderIdDTO = {
  userId: string;
  favoriteArticleFolderId: string;
  offset: number;
  keyword?: string;
};

export const getFavoriteArticlesByFavoriteArticleFolderId = async ({
  userId,
  favoriteArticleFolderId,
  offset,
  keyword,
}: GetFavoriteArticlesByFavoriteArticleFolderIdDTO): Promise<
  Array<FavoriteArticleType>
> => {
  try {
    const supabase = await createGetOnlyServerSideClient();

    const query = supabase
      .from("favorite_articles")
      .select("*")
      .eq("user_id", userId)
      .eq("favorite_article_folder_id", favoriteArticleFolderId);

    if (keyword) {
      query.or(
        `title.ilike.%${keyword}%,description.ilike.%${keyword}%,tags.ilike.%${keyword}%`
      );
    }

    const { data, error } = await query
      .order("created_at", { ascending: false })
      .range((offset - 1) * LIMIT, offset * LIMIT - 1);

    if (error || !data) return [];

    return data.map((favoriteArticle) =>
      convertDatabaseResponseToFavoriteArticleResponse(favoriteArticle)
    );
  } catch (err) {
    throw new Error(`Failed to get favorite articles: ${err}`);
  }
};

type GetFavoriteArticleByIdDTO = {
  id: string;
  userId: string;
};

export const getFavoriteArticleById = async ({
  id,
  userId,
}: GetFavoriteArticleByIdDTO) => {
  try {
    const supabase = await createGetOnlyServerSideClient();

    const query = supabase
      .from("favorite_articles")
      .select("*")
      .eq("id", id)
      .eq("user_id", userId);

    const { data, error } = await query.single();

    if (error || !data) return undefined;

    return convertDatabaseResponseToFavoriteArticleResponse(data);
  } catch (err) {
    throw new Error(`Failed to get favorite article: ${err}`);
  }
};

type GetFavoriteArticleCountByFavoriteArticleFolderIdAndArticleIdAndArticleUrlDTO =
  {
    userId: string;
    articleId: string;
    favoriteArticleFolderId: string;
    articleUrl: string;
  };

export const getFavoriteArticleCountByFavoriteArticleFolderIdAndArticleIdAndArticleUrl =
  async ({
    userId,
    articleId,
    favoriteArticleFolderId,
    articleUrl,
  }: GetFavoriteArticleCountByFavoriteArticleFolderIdAndArticleIdAndArticleUrlDTO): Promise<number> => {
    try {
      const supabase = await createGetOnlyServerSideClient();
      const query = supabase
        .from("favorite_articles")
        .select("*", { count: "exact" })
        .eq("user_id", userId)
        .eq("favorite_article_folder_id", favoriteArticleFolderId)
        .eq("article_id", articleId)
        .eq("article_url", articleUrl);

      const { error, count } = await query;

      if (error || !count) return 0;

      return count;
    } catch (err) {
      throw new Error(`Failed to get favorite article count: ${err}`);
    }
  };

type GetFavoriteArticleCountByFolderIdAndArticleUrlDTO = {
  userId: string;
  favoriteArticleFolderId: string;
  articleUrl: string;
};

export const getFavoriteArticleCountByFolderIdAndArticleUrl = async ({
  userId,
  favoriteArticleFolderId,
  articleUrl,
}: GetFavoriteArticleCountByFolderIdAndArticleUrlDTO) => {
  try {
    const supabase = await createGetOnlyServerSideClient();
    const query = supabase
      .from("favorite_articles")
      .select("*", { count: "exact" })
      .eq("user_id", userId)
      .eq("favorite_article_folder_id", favoriteArticleFolderId)
      .eq("article_url", articleUrl);
    const { error, count } = await query;

    if (error || !count) return 0;

    return count;
  } catch (err) {
    throw new Error(`Failed to get favorite article count: ${err}`);
  }
};

export const getFavoriteArticleCountByFolderIdAndArticleUrlAndArticle = async ({
  userId,
  favoriteArticleFolderId,
  articleUrl,
}: GetFavoriteArticleCountByFolderIdAndArticleUrlDTO) => {
  try {
    const supabase = await createGetOnlyServerSideClient();
    const query = supabase
      .from("favorite_articles")
      .select("*", { count: "exact" })
      .eq("user_id", userId)
      .eq("favorite_article_folder_id", favoriteArticleFolderId)
      .eq("article_url", articleUrl);

    const { error, count } = await query;

    if (error || !count) return 0;

    return count;
  } catch (err) {
    throw new Error(`Failed to get favorite article count: ${err}`);
  }
};

type FavoriteArticleGetDatabaseResponseType =
  Database["public"]["Tables"]["favorite_articles"]["Row"];

const convertDatabaseResponseToFavoriteArticleResponse = (
  favoriteArticle: FavoriteArticleGetDatabaseResponseType
): FavoriteArticleType => {
  return {
    id: favoriteArticle.id,
    favoriteArticleFolderId: favoriteArticle.favorite_article_folder_id,
    platformId: favoriteArticle.platform_id || undefined,
    articleId: favoriteArticle.article_id,
    userId: favoriteArticle.user_id,
    title: favoriteArticle.title,
    description: favoriteArticle.description,
    articleUrl: favoriteArticle.article_url,
    publishedAt: favoriteArticle.published_at || undefined,
    authorName: favoriteArticle.author_name || undefined,
    tags: favoriteArticle.tags || undefined,
    thumbnailUrl: favoriteArticle.thumbnail_url,
    platformName: favoriteArticle.platform_name,
    platformUrl: favoriteArticle.platform_url,
    platformFaviconUrl: favoriteArticle.platform_favicon_url,
    isEng: favoriteArticle.is_eng,
    isRead: favoriteArticle.is_read,
    isPrivate: favoriteArticle.is_private,
    createdAt: favoriteArticle.created_at,
    updatedAt: favoriteArticle.updated_at,
  };
};

/**
 * ==========================================
 * Create
 * ==========================================
 */

export type CreateFavoriteArticleDTO = {
  userId: string;
  favoriteArticleFolderId: string;
  platformId?: string;
  articleId: string;
  title: string;
  description: string;
  articleUrl: string;
  publishedAt?: string;
  authorName?: string;
  tags?: string;
  thumbnailUrl?: string;
  platformName?: string;
  platformUrl?: string;
  platformFaviconUrl?: string;
  isEng?: boolean;
  isRead?: boolean;
  isPrivate?: boolean;
};

export const createFavoriteArticle = async (dto: CreateFavoriteArticleDTO) => {
  try {
    const uuid = uuidv4();
    const supabase = await createGetOnlyServerSideClient();

    const { data, error } = await supabase
      .from("favorite_articles")
      .insert({
        id: uuid,
        user_id: dto.userId,
        favorite_article_folder_id: dto.favoriteArticleFolderId,
        platform_id: dto.platformId,
        article_id: dto.articleId,
        title: dto.title,
        description: dto.description,
        article_url: dto.articleUrl,
        published_at: dto.publishedAt,
        author_name: dto.authorName,
        tags: dto.tags,
        thumbnail_url: dto.thumbnailUrl || "",
        platform_name: dto.platformName || "",
        platform_url: dto.platformUrl || "",
        platform_favicon_url: dto.platformFaviconUrl || "",
        is_eng: dto.isEng,
        is_read: dto.isRead,
        is_private: dto.isPrivate,
      })
      .select();

    if (error || !data) return;
    return convertDatabaseResponseToFavoriteArticleResponse(data[0]);
  } catch (err) {
    throw new Error(`Failed to create favorite article: ${err}`);
  }
};

/**
 * ==========================================
 * Delete
 * ==========================================
 */

type DeleteFavoriteArticleDTO = {
  id: string;
  userId: string;
};

export const deleteFavoriteArticle = async ({
  id,
  userId,
}: DeleteFavoriteArticleDTO) => {
  try {
    const supabase = await createGetOnlyServerSideClient();
    const { error } = await supabase
      .from("favorite_articles")
      .delete()
      .eq("id", id)
      .eq("user_id", userId);
    if (error) return;
    return id;
  } catch (err) {
    throw new Error(`Failed to delete favorite article: ${err}`);
  }
};

File: micro-service/favorite-service/internal/adapter/persistence_adapter/favorite_article.go
package persistenceadapter

import (
	"context"
	"fmt"

	fpb "github.com/YukiOnishi1129/checkpicks-protocol-buffers/checkpicks-rpc-go/grpc/favorite"
	"github.com/YukiOnishi1129/techpicks/micro-service/favorite-service/internal/domain/entity"
	"github.com/YukiOnishi1129/techpicks/micro-service/favorite-service/internal/domain/repository"
	"github.com/google/uuid"
	"github.com/volatiletech/null/v8"
	"github.com/volatiletech/sqlboiler/v4/queries/qm"
)

type FavoriteArticlePersistenceAdapter interface {
	GetFavoriteArticles(ctx context.Context, req *fpb.GetFavoriteArticlesRequest, limit int) (entity.FavoriteArticleSlice, error)
	GetFavoriteArticlesOrderByArticleID(ctx context.Context, req *fpb.GetFavoriteAllFolderArticlesRequest, cursor string, limit int) (entity.FavoriteArticleSlice, error)
	GetFavoriteArticlesByArticleID(ctx context.Context, articleID, userID string) (entity.FavoriteArticleSlice, error)
	GetFavoriteArticleByArticleIDAndFavoriteArticleFolderID(ctx context.Context, articleID, favoriteArticleFolderID, userID string) (entity.FavoriteArticle, error)
	GetFavoriteArticlesByFavoriteArticleFolderID(ctx context.Context, fafID, userID string, limit *int, isAllFetch *bool) (entity.FavoriteArticleSlice, error)
	GetFavoriteArticleByID(ctx context.Context, id string, userID string) (entity.FavoriteArticle, error)
	GetFavoriteArticleByArticleURL(ctx context.Context, articleURL, favoriteArticleFolderID, userID string) (entity.FavoriteArticle, error)
	CountPreviousFavoriteArticleByArticleID(ctx context.Context, id, articleID, userID string) (int64, error)

	CreateFavoriteArticle(ctx context.Context, req *fpb.CreateFavoriteArticleRequest) (entity.FavoriteArticle, error)
	CreateFavoriteArticleForUploadArticle(ctx context.Context, dto CreateFavoriteArticleForUploadArticleDTO) (entity.FavoriteArticle, error)
	DeleteFavoriteArticle(ctx context.Context, fa entity.FavoriteArticle) error
	MultiDeleteFavoriteArticles(ctx context.Context, fa entity.FavoriteArticleSlice) error
}

type favoriteArticlePersistenceAdapter struct {
	favoriteArticleRepository repository.FavoriteArticleRepository
}

func NewFavoriteArticlePersistenceAdapter(favoriteArticleRepository repository.FavoriteArticleRepository) FavoriteArticlePersistenceAdapter {
	return &favoriteArticlePersistenceAdapter{
		favoriteArticleRepository: favoriteArticleRepository,
	}
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticles(ctx context.Context, req *fpb.GetFavoriteArticlesRequest, limit int) (entity.FavoriteArticleSlice, error) {
	q := []qm.QueryMod{
		qm.Where("user_id = ?", req.GetUserId()),
		qm.OrderBy("created_at DESC"),
		qm.Limit(limit),
	}
	if req.GetCursor() != nil {
		q = append(q, qm.Where("created_at < (SELECT created_at FROM favorite_articles WHERE id = ?)", req.GetCursor().GetValue()))
	}
	if req.GetKeywords() != nil {
		for _, keyword := range req.GetKeywords() {
			q = append(q, qm.Expr(
				qm.And("title ILIKE ?", "%"+keyword.GetValue()+"%"),
				qm.Or("description ILIKE ?", "%"+keyword.GetValue()+"%"),
			))
		}
	}
	if req.GetFavoriteArticleFolderId() != nil {
		q = append(q, qm.Where("favorite_article_folder_id = ?", req.GetFavoriteArticleFolderId().GetValue()))
	}

	favoriteArticles, err := fapa.favoriteArticleRepository.GetFavoriteArticles(ctx, q)
	if err != nil {
		return nil, err
	}
	return favoriteArticles, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticlesOrderByArticleID(ctx context.Context, req *fpb.GetFavoriteAllFolderArticlesRequest, cursor string, limit int) (entity.FavoriteArticleSlice, error) {
	q := []qm.QueryMod{
		qm.Where("user_id = ?", req.GetUserId()),
		qm.OrderBy("created_at DESC"),
		qm.OrderBy("article_id"),
		qm.Limit(limit),
	}

	if cursor != "" {
		q = append(q, qm.Where("created_at < (SELECT created_at FROM favorite_articles WHERE id = ?)", cursor))
	}

	if req.GetKeywords() != nil {
		for _, keyword := range req.GetKeywords() {
			q = append(q, qm.Expr(
				qm.And("title LIKE ?", "%"+keyword.GetValue()+"%"),
				qm.Or("description LIKE ?", "%"+keyword.GetValue()+"%"),
			))
		}
	}

	favoriteArticles, err := fapa.favoriteArticleRepository.GetFavoriteArticles(ctx, q)
	if err != nil {
		fmt.Printf("Error executing query: %v\n", err)
		return nil, err
	}

	return favoriteArticles, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticlesByArticleID(ctx context.Context, articleID, userID string) (entity.FavoriteArticleSlice, error) {
	q := []qm.QueryMod{
		qm.Where("article_id = ?", articleID),
		qm.Where("user_id = ?", userID),
		qm.Load(qm.Rels(entity.FavoriteArticleRels.FavoriteArticleFolder)),
	}
	fa, err := fapa.favoriteArticleRepository.GetFavoriteArticles(ctx, q)
	if err != nil {
		return entity.FavoriteArticleSlice{}, err
	}
	return fa, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticleByArticleIDAndFavoriteArticleFolderID(ctx context.Context, articleID, favoriteArticleFolderID, userID string) (entity.FavoriteArticle, error) {
	q := []qm.QueryMod{
		qm.Where("article_id = ?", articleID),
		qm.Where("favorite_article_folder_id = ?", favoriteArticleFolderID),
		qm.Where("user_id = ?", userID),
	}
	fa, err := fapa.favoriteArticleRepository.GetFavoriteArticle(ctx, q)
	if err != nil {
		return entity.FavoriteArticle{}, err
	}
	return fa, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticleByID(ctx context.Context, id string, userID string) (entity.FavoriteArticle, error) {
	q := []qm.QueryMod{
		qm.Where("user_id = ?", userID),
	}
	fa, err := fapa.favoriteArticleRepository.GetFavoriteArticleByID(ctx, id, q)
	if err != nil {
		return entity.FavoriteArticle{}, err
	}
	return fa, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticlesByFavoriteArticleFolderID(ctx context.Context, fafID, userID string, limit *int, isFavoriteArticleAllFetch *bool) (entity.FavoriteArticleSlice, error) {

	q := []qm.QueryMod{
		qm.Where("favorite_article_folder_id = ?", fafID),
		qm.Where("user_id = ?", userID),
		qm.OrderBy("created_at DESC"),
	}

	if isFavoriteArticleAllFetch == nil || !*isFavoriteArticleAllFetch {
		paramLimit := 1
		if limit != nil {
			paramLimit = int(*limit)
		}
		q = append(q, qm.Limit(paramLimit))
	}

	favoriteArticles, err := fapa.favoriteArticleRepository.GetFavoriteArticles(ctx, q)
	if err != nil {
		return nil, err
	}
	return favoriteArticles, nil
}

func (fapa *favoriteArticlePersistenceAdapter) GetFavoriteArticleByArticleURL(ctx context.Context, articleURL, favoriteArticleFolderID, userID string) (entity.FavoriteArticle, error) {
	q := []qm.QueryMod{
		qm.Where("article_url = ?", articleURL),
		qm.Where("user_id = ?", userID),
		qm.Where("favorite_article_folder_id = ?", favoriteArticleFolderID),
	}
	fa, err := fapa.favoriteArticleRepository.GetFavoriteArticle(ctx, q)
	if err != nil {
		return entity.FavoriteArticle{}, err
	}
	return fa, nil
}

func (fapa *favoriteArticlePersistenceAdapter) CountPreviousFavoriteArticleByArticleID(ctx context.Context, id, articleID, userID string) (int64, error) {
	q := []qm.QueryMod{
		qm.Where("article_id = ?", articleID),
		qm.Where("user_id = ?", userID),
		qm.Where("created_at > (SELECT created_at FROM favorite_articles WHERE id = ?)", id),
	}
	count, err := fapa.favoriteArticleRepository.CountFavoriteArticles(ctx, q)
	if err != nil {
		return 0, err
	}
	return count, nil
}

func (fapa *favoriteArticlePersistenceAdapter) CreateFavoriteArticle(ctx context.Context, req *fpb.CreateFavoriteArticleRequest) (entity.FavoriteArticle, error) {
	favoriteArticleID, _ := uuid.NewRandom()
	fa := entity.FavoriteArticle{
		ID:                      favoriteArticleID.String(),
		UserID:                  req.GetUserId(),
		FavoriteArticleFolderID: req.GetFavoriteArticleFolderId(),
		ArticleID:               req.GetArticleId(),
		Title:                   req.GetTitle(),
		ArticleURL:              req.GetArticleUrl(),
		ThumbnailURL:            req.GetThumbnailUrl(),
		PlatformName:            req.GetPlatformName(),
		PlatformURL:             req.GetPlatformUrl(),
		PlatformFaviconURL:      req.GetPlatformFaviconUrl(),
		IsEng:                   req.GetIsEng(),
		IsPrivate:               req.GetIsPrivate(),
	}

	if req.GetDescription() != nil {
		fa.Description = req.GetDescription().GetValue()
	}
	if req.GetPlatformId() != nil {
		fa.PlatformID = null.String{String: req.GetPlatformId().GetValue(), Valid: true}
	}
	if req.GetPublishedAt() != nil {
		fa.PublishedAt = null.TimeFrom(req.GetPublishedAt().AsTime())
	}
	if req.GetAuthorName() != nil {
		fa.AuthorName = null.String{String: req.GetAuthorName().GetValue(), Valid: true}
	}
	if req.GetTags() != nil {
		fa.Tags = null.String{String: req.GetTags().GetValue(), Valid: true}
	}

	if _, err := fapa.favoriteArticleRepository.CreateFavoriteArticle(ctx, fa); err != nil {
		return entity.FavoriteArticle{}, err
	}
	return fa, nil
}

type CreateFavoriteArticleForUploadArticleDTO struct {
	UserID                  string
	FavoriteArticleFolderID string
	PlatformID              null.String
	ArticleID               string
	Title                   string
	Description             string
	ArticleURL              string
	PublishedAt             null.Time
	AuthorName              null.String
	Tags                    null.String
	ThumbnailURL            string
	PlatformName            string
	PlatformURL             string
	PlatformFaviconURL      string
	IsEng                   bool
	IsPrivate               bool
}

func (fapa *favoriteArticlePersistenceAdapter) CreateFavoriteArticleForUploadArticle(ctx context.Context, dto CreateFavoriteArticleForUploadArticleDTO) (entity.FavoriteArticle, error) {
	favoriteArticleID, _ := uuid.NewRandom()
	fa := entity.FavoriteArticle{
		ID:                      favoriteArticleID.String(),
		UserID:                  dto.UserID,
		FavoriteArticleFolderID: dto.FavoriteArticleFolderID,
		ArticleID:               dto.ArticleID,
		Title:                   dto.Title,
		Description:             dto.Description,
		ArticleURL:              dto.ArticleURL,
		ThumbnailURL:            dto.ThumbnailURL,
		PlatformName:            dto.PlatformName,
		PlatformURL:             dto.PlatformURL,
		PlatformFaviconURL:      dto.PlatformFaviconURL,
		IsEng:                   dto.IsEng,
		IsPrivate:               dto.IsPrivate,
		IsRead:                  false,
	}

	if dto.PlatformID.Valid {
		fa.PlatformID = dto.PlatformID
	}
	if dto.PublishedAt.Valid {
		fa.PublishedAt = dto.PublishedAt
	}
	if dto.AuthorName.Valid {
		fa.AuthorName = dto.AuthorName
	}
	if dto.Tags.Valid {
		fa.Tags = dto.Tags
	}

	if _, err := fapa.favoriteArticleRepository.CreateFavoriteArticle(ctx, fa); err != nil {
		return entity.FavoriteArticle{}, err
	}
	return fa, nil
}

func (fapa *favoriteArticlePersistenceAdapter) DeleteFavoriteArticle(ctx context.Context, fa entity.FavoriteArticle) error {
	return fapa.favoriteArticleRepository.DeleteFavoriteArticle(ctx, fa)
}

func (fapa *favoriteArticlePersistenceAdapter) MultiDeleteFavoriteArticles(ctx context.Context, fa entity.FavoriteArticleSlice) error {
	return fapa.favoriteArticleRepository.MultiDeleteFavoriteArticles(ctx, fa)
}
